### DISCUSSION TITLE
Swift Binding issues?

### DISCUSSION THREAD

**Discussion by JarretHoving97** [pointfreeco/swift-composable-architecture#2304] [Q&A] [answered]
Hi there!

I'm currently trying to figure out TCA in SwiftUI and struggling with ViewBInding that does not work correctly. 

On simulator, the binding will not work at all. On actual de device it works MOST of the time. and on Canvas it runs just fine. 
I think there could be an issue in my StackBasedNavigation logic or it could be a overall bug. 

I would really like to know what's wrong either way.

I've also tried viewStore.bind() and that also did not seem to work. 

Here's my repo
https://github.com/JarretHoving97/binding-tca-issue

Views / reducers
https://github.com/JarretHoving97/binding-tca-issue/tree/main/dinner-cycle/dinner-cycle/Scene/Recipe/RecipeCategory

https://github.com/JarretHoving97/binding-tca-issue/tree/main/dinner-cycle/dinner-cycle/Scene/Recipe/NewRecipe

**Comment by oliverfoggin** [comment]
Please can you link to just the file (or view/reducer) with the issue in it.

It's hard to find where the issue is as that project has quite a large directory structure.

Thanks

**Reply by JarretHoving97** [reply]
Added links to view/reducers!

**Comment by mbrandonw** [comment]
Hi @JarretHoving97, this project does not build due to this error:

> 🛑 Build input file cannot be found: 'binding-tca-issue-main/dinner-cycle/dinner-cycle/Info.plist'. Did you forget to declare this file as an output of a script phase or custom build rule which produces it?

However, I will say that this project is much too big to diagnose. It would be better if you can create a small sample project from scratch with the minimal number of pieces that demonstrates the problem. A benefit of doing that is along the way you might figure out the bug yourself. 🙂

**Reply by JarretHoving97** [reply]
Hi Brandon, thanks for looking! 

Oh, my bad about the info.plist.

I tried to make the project smaller. (Only view + reducer) but the problem
Will not reproduce.

Propably something to do with how the project is layered thats why i pushed the whole project instead. 

I will add the info.plist file and build the project step by step again and let it know here when i find the issue!

**Comment by JarretHoving97** [comment, marked as answer]
Okay.. So started removing every binding from the project and started at the rootView (thankfully) where I noticed I did some unnecessary binding. 

`struct RootView: View {
    
    let store: StoreOf<RootViewStore>
    
    var body: some View {
        
        WithViewStore(self.store, observe: {$0}) { viewStore in
            TabView(
                selection: viewStore.binding(
                    get: \.selected,
                    send: RootViewStore.Action.selectTab)
            )
     ...
`

Has been changed to:

`struct RootView: View {
    
    let store: StoreOf<RootViewStore>
    
    var body: some View {
        
        WithViewStore(self.store, observe: {$0}) { viewStore in
            TabView {
                RecipeView(
                    store: Store(
                        initialState: .init(),
                        reducer: RecipeViewStore(client: .live)
                    )
                )
      ...
`

@BindingState values for textfields gets updated as it should be.

**Reply by mbrandonw** [reply]
Glad you figured it out, but you should also be careful to not observe too much state. If you aren't using a `viewStore` then there is no need to use `WithViewStore`, and if you only need a single piece of state from the view store then you should only observe that state.

We have an article about this:

https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/performance#View-stores
